### PR TITLE
Iteration-aware trigger

### PR DESCRIPTION
**Pull request by niboshi**
Related: #5585

### CODE REVIEW

**Comment by niboshi** [comment]
This PR is blocked by #5645.

**Comment by niboshi** [comment]
PTAL

**Comment by niboshi** [comment]
>cached_through_iteration

Doesn't that sound like the cache lives across iterations?
Perhaps `cached_during_iteration`? (I tempted to include `each` or `every`, but actually the cache does not necessarily exist in every iteration).

**Comment by hvy** [comment]
I have not followed the discussion behind this PR so excuse me if this is nonsense. May I ask why this issue was generalized to all triggers? It sounds like a bug in only some of the triggers incl. best value trigger to me? 🤔  IMHO the concept of "iteration" is necessarily inherent to all triggers (e.g. the `TimeTrigger`) and this PR sort of introduces an assumption that a trigger are always evaluated on an iteration basis. Although true, and I understand it from a practical point of view, but am curious from a design perspective. I guess one way to see it would be to have a `Trigger` class that doesn't care about iterations, and a `IterationAwareTrigger(Trigger)`, instead of a flag that is?

**Comment by niboshi** [comment]
@hvy All triggers receive a trainer. I think it's natural that the concepts of triggers and iterations are coupled.

**Comment by hvy** [comment]
For the record, discussed offline and agreed that this flag is `cached_during_iteration` is not needed. Rather we should make sure all triggers behave as if "iteration aware" by default.

File: chainer/training/trigger.py [hunk 1]
```diff
@@ -1,4 +1,146 @@
...
+        @chainer.trigger
+        def random_trigger(trainer):
+            # A trigger that triggers randomly at 10% of all the iterations.
+            return random.randint(0, 9) == 0
+
+    See also:
+        :func:`chainer.trigger`
+    """
+
+    cached_during_iteration = True
```
Comment (hvy) [inline chainer/training/trigger.py]: Agreed to remove this flag as per earlier comment.

File: chainer/training/trigger.py [hunk 2]
```diff
@@ -1,4 +1,146 @@
...
+
+    # None if the result is NOT cached during an iteration.
+    # Otherwise a tuple containing two elements:
+    # [0] epoch detail and [1] the evaluation (True/False).
+    __cache_during_iteration = None
+
+    def evaluate(self, trainer):
```
[the remark this answers is not in the document]
Comment (niboshi) [reply chainer/training/trigger.py]: Hmm, I don't think `trigger.evaluate()` is not abstract at all.

File: chainer/training/trigger.py (hunk 1 above)
Comment (niboshi) [reply chainer/training/trigger.py]: I don't remember the argument, but should `TimeTrigger` for example be also iteration-aware?

**Comment by Saanidhyavats** [comment]
Whether someone else is working on this issue or shall I start contributing on it

**Pull request closed without merging**